### ISSUE TITLE
Investigate custom JIterableWrapper for joins & cogroups

### ISSUE THREAD

**Issue by nevillelyh** [spotify/scio#891]
`Iterable#map` is eager due to the implicit `CBF`:
`def map[B, That](f: A => B)(implicit bf: CanBuildFrom[Repr, B, That]): That`
So doing `xs.map` in a `cogroup` may OOM. Not sure there's anything we can do though without changing the type of groups from `Iterable[V]` to something else.

There might be something we can do though for join/outer join of 2-4 way cases, e.g. handcraft loops and avoid conversion instead of the script generated for-comprehension. Also skipping `flatMap` and access `ProcessContext#output` directly. These 2 together should give us some speed up?

```scala
val (key, result) = (kv.getKey, kv.getValue)
for {
  b <- result.getAll(tagB).asScala.iterator
  a <- result.getAll(tagA).asScala.iterator
} yield (key, (a, b))
```

**Comment by nevillelyh** [comment]
Pushed some WIP & benchmark here:
https://github.com/spotify/scio/blob/neville/join/scio-jmh/src/test/scala/com/spotify/scio/jmh/JoinBenchmark.scala

```
[info] Benchmark                Mode  Cnt      Score     Error  Units
[info] JoinBenchmark.artisanA   avgt  200     42.116 ±   2.458  ns/op
[info] JoinBenchmark.artisanB   avgt  200     39.517 ±   3.531  ns/op
[info] JoinBenchmark.artisanC   avgt  200      7.573 ±   0.187  ns/op
[info] JoinBenchmark.artisanD   avgt  200      7.820 ±   0.232  ns/op
[info] JoinBenchmark.forYieldA  avgt  200    178.239 ±   5.256  ns/op
[info] JoinBenchmark.forYieldB  avgt  200    672.130 ±  41.797  ns/op
[info] JoinBenchmark.forYieldC  avgt  200  10605.416 ± 316.775  ns/op
[info] JoinBenchmark.forYieldD  avgt  200  16340.696 ± 701.269  ns/op
```

**Comment by nevillelyh** [comment]
Fixed in #901
